### PR TITLE
Feature 303 metaform cacheability update

### PR DESCRIPTION
**Pull request by mrhaekkinen**
Statistics calculation update

### CODE REVIEW

File: src/main/kotlin/fi/metatavu/metaform/server/controllers/ReplyController.kt [hunk 1]
```diff
@@ -171,6 +188,12 @@ class ReplyController {
      */
     fun updateResourceId(reply: Reply, resourceId: UUID?): Reply {
         replyDAO.updateResourceId(reply, resourceId)
+        replyUpdatedEvent.fire(
+                ReplyUpdatedEvent(
+                        replyId = reply.id!!,
+                        metaformId = reply.metaform.id!!
+                )
+        )
```
Comment (villejuutila) [inline src/main/kotlin/fi/metatavu/metaform/server/controllers/ReplyController.kt:196]: Is this needed?

File: src/main/kotlin/fi/metatavu/metaform/server/rest/RepliesApi.kt [hunk 2]
```diff
@@ -237,6 +240,7 @@ class RepliesApi: fi.metatavu.metaform.api.spec.RepliesApi, AbstractApi() {
     auditLogEntryController.generateAuditLog(metaform, userId, reply.id!!, null, null, AuditLogEntryType.DELETE_REPLY)
     replyController.deleteReply(reply)
 
+
```
Comment (villejuutila) [inline src/main/kotlin/fi/metatavu/metaform/server/rest/RepliesApi.kt:243]: Move ReplyEvent triggering to here from the controller.

File: src/main/kotlin/fi/metatavu/metaform/server/controllers/ReplyController.kt [hunk 3]
```diff
@@ -120,15 +130,22 @@ class ReplyController {
...
+        val createdReply = replyDAO.create(
+                id = id,
+                userId = userId,
+                metaform = metaform,
+                resourceId = null,
+                privateKey = privateKey?.encoded,
+                revision = null,
+                lastModifierId = loggedUserId
         )
+        replyCreatedEvent.fire(
```
Comment (villejuutila) [inline src/main/kotlin/fi/metatavu/metaform/server/controllers/ReplyController.kt:142]: Move this to implementation layer.

**Review by villejuutila — CHANGES_REQUESTED** [review]

File: src/main/kotlin/fi/metatavu/metaform/server/controllers/MetaformController.kt [hunk 4]
```diff
@@ -257,14 +257,14 @@ class MetaformController {
...
-                        keycloakClient,
-                    reply.resourceId ?: throw ResourceNotFoundException("Resource not found"),
-                        resourceName,
-                        listOf(AuthorizationScope.REPLY_NOTIFY)
-                )
+            if (replyCreated) emptySet()
+            else metaformKeycloakController.getResourcePermittedUsers(
+                    adminClient,
+                    keycloakClient,
+                reply.resourceId ?: throw ResourceNotFoundException("Resource not found"),
```
[the remark this answers is not in the document]
Comment (mrhaekkinen) [reply src/main/kotlin/fi/metatavu/metaform/server/controllers/MetaformController.kt:264]: Changed to named argument syntax and reformatted.

File: src/main/kotlin/fi/metatavu/metaform/server/controllers/ReplyController.kt (hunk 1 above)
Comment (mrhaekkinen) [reply src/main/kotlin/fi/metatavu/metaform/server/controllers/ReplyController.kt:196]: Actually seems not, removing.

File: src/main/kotlin/fi/metatavu/metaform/server/controllers/MetaformController.kt [hunk 5]
```diff
@@ -26,6 +25,7 @@ import org.keycloak.admin.client.resource.UserResource
 import org.keycloak.representations.idm.UserRepresentation
 import java.util.*
 import javax.enterprise.context.ApplicationScoped
+import javax.enterprise.event.Event
```
[the remark this answers is not in the document]
Comment (mrhaekkinen) [reply src/main/kotlin/fi/metatavu/metaform/server/controllers/MetaformController.kt:28]: Removed.

File: src/main/kotlin/fi/metatavu/metaform/server/rest/RepliesApi.kt (hunk 2 above)
Comment (mrhaekkinen) [reply src/main/kotlin/fi/metatavu/metaform/server/rest/RepliesApi.kt:243]: As per discussed, did this change in order to keep the logic intact.

**Review by anttileppa — CHANGES_REQUESTED** [review]

File: src/main/kotlin/fi/metatavu/metaform/server/controllers/ReplyEvent.kt [hunk 6]
```diff
@@ -0,0 +1,11 @@
+package fi.metatavu.metaform.server.controllers
+
+import java.util.*
+
+/**
+ * @param metaformId Metaform id
+ * @param replyId Reply id
+  */
+data class ReplyCreatedEvent(val metaformId: UUID, val replyId: UUID)
```
[the remark this answers is not in the document]
Comment (mrhaekkinen) [reply src/main/kotlin/fi/metatavu/metaform/server/controllers/ReplyEvent.kt:9]: Classes separated to them own files.

File: src/main/kotlin/fi/metatavu/metaform/server/controllers/ReplyController.kt [hunk 7]
```diff
@@ -1008,6 +1037,30 @@ class ReplyController {
         }
     }
 
+    /**
+     * Triggers ReplyEvent depending on given boolean value
+     *
+     * @param reply Reply
+     * @param replyCreated Boolean
+     */
+    fun triggerReplyEvent(reply: Reply, replyCreated: Boolean) {
```
[the remark this answers is not in the document]
Comment (mrhaekkinen) [reply src/main/kotlin/fi/metatavu/metaform/server/controllers/ReplyController.kt:1046]: Separate functions implemented.

File: src/main/kotlin/fi/metatavu/metaform/server/controllers/ReplyController.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (mrhaekkinen) [reply src/main/kotlin/fi/metatavu/metaform/server/controllers/ReplyController.kt:142]: Yes indeed. This is now prevented.

File: src/main/kotlin/fi/metatavu/metaform/server/controllers/MetaformStatisticsObserver.kt [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+        onReplyEvent(event.metaformId)
+    }
+
+    /**
+     * Finds Metaform by given UUID and triggers statistics recalculation for given Metaform after successful
+     * create/delete/update event
+     *
+     * @param metaformId
+     */
+    private fun onReplyEvent(metaformId: UUID) {
```
[the remark this answers is not in the document]
Comment (mrhaekkinen) [reply src/main/kotlin/fi/metatavu/metaform/server/controllers/MetaformStatisticsObserver.kt:69]: Done.

File: src/main/kotlin/fi/metatavu/metaform/server/controllers/ReplyController.kt [hunk 9]
```diff
@@ -346,6 +369,12 @@ class ReplyController {
     fun deleteReply(reply: Reply) {
         anyReplyFieldDAO.listByReply(reply).forEach { deleteField(it) }
         replyDAO.delete(reply)
+        replyDeletedEvent.fire(
```
[the remark this answers is not in the document]
Comment (mrhaekkinen) [reply src/main/kotlin/fi/metatavu/metaform/server/controllers/ReplyController.kt:372]: Done.

File: src/main/kotlin/fi/metatavu/metaform/server/controllers/ReplyController.kt [hunk 10]
```diff
@@ -171,6 +188,12 @@ class ReplyController {
      */
     fun updateResourceId(reply: Reply, resourceId: UUID?): Reply {
         replyDAO.updateResourceId(reply, resourceId)
+        replyUpdatedEvent.fire(
```
[the remark this answers is not in the document]
Comment (mrhaekkinen) [reply src/main/kotlin/fi/metatavu/metaform/server/controllers/ReplyController.kt:191]: Fixed this.

File: src/main/kotlin/fi/metatavu/metaform/server/controllers/ReplyController.kt (hunk 3 above)
Comment (mrhaekkinen) [reply src/main/kotlin/fi/metatavu/metaform/server/controllers/ReplyController.kt:142]: Done

File: src/main/kotlin/fi/metatavu/metaform/server/controllers/MetaformStatisticsObserver.kt [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+import javax.annotation.PostConstruct
+import javax.enterprise.context.ApplicationScoped
+import javax.enterprise.event.Observes
+import javax.enterprise.event.TransactionPhase
+import javax.inject.Inject
+import javax.transaction.Transactional
+
+@ApplicationScoped
+@Suppress ("unused")
+class MetaformStatisticsObserver {
```
Comment (anttileppa) [inline src/main/kotlin/fi/metatavu/metaform/server/controllers/MetaformStatisticsObserver.kt:14]: Add class docs

**Review by anttileppa — APPROVED** [review]

**Review by villejuutila — APPROVED** [review]

**Pull request merged**